x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/PostAnalyzer.java

```diff
@@ -47,21 +49,24 @@ public LogicalPlan postAnalyze(LogicalPlan plan, EqlConfiguration configuration)
             Holder<Boolean> hasJoin = new Holder<>(Boolean.FALSE);
 
             Source projectCtx = synthetic("<implicit-project>");
-            // first per KeyedFilter
-            plan = plan.transformUp(KeyedFilter.class, k -> {
+            if (plan.anyMatch(Sequence.class::isInstance)) {
                 hasJoin.set(Boolean.TRUE);
-                Project p = new Project(projectCtx, k.child(), k.extractionAttributes());
+                // first per KeyedFilter
+                plan = plan.transformUp(KeyedFilter.class, k -> {
+                    Project p = new Project(projectCtx, k.child(), k.extractionAttributes());
 
-                // TODO: this could be incorporated into the query generation
-                LogicalPlan fetchSize = new LimitWithOffset(
-                    synthetic("<fetch-size>"),
-                    new Literal(synthetic("<fetch-value>"), configuration.fetchSize(), DataTypes.INTEGER),
-                    p
-                );
+                    // TODO: this could be incorporated into the query generation
+                    LogicalPlan fetchSize = new LimitWithOffset(
+                        synthetic("<fetch-size>"),
+                        new Literal(synthetic("<fetch-value>"), configuration.fetchSize(), DataTypes.INTEGER),
+                        p
+                    );
 
-                return new KeyedFilter(k.source(), fetchSize, k.keys(), k.timestamp(), k.tiebreaker());
-            });
+                    return new KeyedFilter(k.source(), fetchSize, k.keys(), k.timestamp(), k.tiebreaker());
+                });
+            }
 
+            hasJoin.set(hasJoin.get() || plan.anyMatch(Sample.class::isInstance));
```

Comment:
I've done it the way it is in the PR because the first `anyMatch` is already traversing the tree and I wanted to take advantage of that step and not do the traversal twice in all cases.